presto-main/src/main/java/com/facebook/presto/server/JavaVersion.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.StandardSystemProperty;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JavaVersion
+{
+    // As described in JEP-223
+    private static final String VERSION_NUMBER = "(?<MAJOR>[1-9][0-9]*)(\\.(?<MINOR>(0|[1-9][0-9]*))(\\.(?<SECURITY>(0|[1-9][0-9]*)))?)?";
+    private static final String PRE = "(?:-(?<PRE>[a-zA-Z0-9]+))?";
+    private static final String BUILD = "(?:(?<PLUS>\\+)(?<BUILD>0|[1-9][0-9]*)?)?";
+    private static final String OPT = "(?:-(?<OPT>[-a-zA-Z0-9.]+))?";
+    private static final Pattern PATTERN = Pattern.compile("^" + VERSION_NUMBER + PRE + BUILD + OPT + "$");
+
+    // For Java 8 and below
+    private static final Pattern LEGACY_PATTERN = Pattern.compile("^1\\.(?<MAJOR>[0-9]+)(\\.(?<MINOR>(0|[1-9][0-9]*)))?(_(?<UPDATE>[1-9][0-9]*))?$");
```

Comment:
Yes, you're right, since we're using `matches()` below. In any case, it's consistent with `PATTERN` above.